### PR TITLE
[NPU]MindSpore backend support eagle3

### PR DESCRIPTION
**Pull request by zzzzzzzxh**
## Motivation
This PR enables EAGLE speculative decoding support for the MindSpore backend ( sgl-mindspore ).

During the integration, we identified a critical issue in EagleWorkerV2 where the input_ids passed to the draft model during the second draft step (immediately following the Extend phase) were incorrect due to a mismatch in how topk_index was processed. This PR fixes that data flow to ensure the Draft model receives the correct tokens generated by the Extend phase.

This PR provides the necessary frontend/worker-side logic and is designed to work in conjunction with mindspore-lab/sgl-mindspore#19 , which implements the backend model definitions ( LlamaEagleModel , LlamaAttention modifications, etc.).

## Modifications

1. Expose hot_token_id Property :

- Added a hot_token_id property to the model runner to retrieve and convert the EAGLE model's tree structure (hot tokens) from MindSpore tensor to PyTorch tensor, enabling the worker to access the draft model's token map.

2. Refine Prefill Logic ( get_is_prefill ) :

- Implemented a dedicated get_is_prefill method to correctly distinguish between standard prefill (Extend) and speculative phases ( Draft Extend , Target Verify ).
- Ensured that speculative decoding phases (like is_draft_extend_v2 , is_target_verify ) are not treated as standard prefill. This forces the backend to use PagedAttention (with KV cache) instead of FlashAttentionScore, which is critical for correctly handling the KV cache during speculative steps.

3. Support EAGLE Verify Step in prepare_inputs :

- Updated input preparation logic to handle ForwardMode.TARGET_VERIFY .
- Adjusted batch_valid_length and q_seq_lens calculation: When in verification mode, it now correctly accounts for the multiple speculative tokens (tree leaves) by adding/multiplying with forward_batch.spec_info.num_tokens_per_batch . This ensures the MindSpore model receives the correct sequence lengths for the verification batch.

## Accuracy Tests

gsm8k 
qwen3 32B eagle3  2p
```
    --speculative-num-steps 3 \
    --speculative-eagle-topk 1 \
    --speculative-num-draft-tokens 3

python3 bench_sglang.py --num-questions 500 --parallel 8
```
```
mindspore:
Accuracy: 0.868
Invalid: 0.000
Latency: 721.831 s
Output throughput: 185.070 token/s


without eagle3:
Accuracy: 0.862
Invalid: 0.000
Latency: 772.881 s
Output throughput: 173.799 token/s

```

## Benchmarking and Profiling
```
    --speculative-num-steps 4 \
    --speculative-eagle-topk 1 \
    --speculative-num-draft-tokens 5
python -m sglang.bench_serving --backend sglang --model /home/Qwen3-8B --num-prompts 50 --dataset-path /home/wangtiance/ShareGPT_V3_unfiltered_cleaned_split.json --max-concurrency 8
```
```
mindspore:
============ Serving Benchmark Result ============
Backend:                                 sglang    
Traffic request rate:                    inf       
Max request concurrency:                 8         
Successful requests:                     50        
Benchmark duration (s):                  128.43    
Total input tokens:                      26955     
Total input text tokens:                 26955     
Total generated tokens:                  23606     
Total generated tokens (retokenized):    23606     
Request throughput (req/s):              0.39      
Input token throughput (tok/s):          209.88    
Output token throughput (tok/s):         183.80    
Peak output token throughput (tok/s):    299.00    
Peak concurrent requests:                10        
Total token throughput (tok/s):          393.68    
Concurrency:                             7.01      
Accept length:                           2.14      
----------------End-to-End Latency----------------
Mean E2E Latency (ms):                   18001.81  
Median E2E Latency (ms):                 14940.41  
P90 E2E Latency (ms):                    38635.95  
P99 E2E Latency (ms):                    46880.54  
---------------Time to First Token----------------
Mean TTFT (ms):                          412.16    
Median TTFT (ms):                        295.32    
P99 TTFT (ms):                           1080.79   
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          36.49     
Median TPOT (ms):                        32.99     
P99 TPOT (ms):                           71.72     
---------------Inter-Token Latency----------------
Mean ITL (ms):                           37.34     
Median ITL (ms):                         25.61     
P95 ITL (ms):                            78.27     
P99 ITL (ms):                            123.28    
Max ITL (ms):                            325.89    
==================================================
```
```
without eagle3
============ Serving Benchmark Result ============						
Backend:                                 sglang    						
Traffic request rate:                    inf       						
Max request concurrency:                 8         						
Successful requests:                     50        						
Benchmark duration (s):                  150.82    						
Total input tokens:                      26955     						
Total input text tokens:                 26955     						
Total generated tokens:                  23606     						
Total generated tokens (retokenized):    23606     						
Request throughput (req/s):              0.33      						
Input token throughput (tok/s):          178.73    						
Output token throughput (tok/s):         156.52    						
Peak output token throughput (tok/s):    192.00    						
Peak concurrent requests:                10        						
Total token throughput (tok/s):          335.25    						
Concurrency:                             7.17      						
----------------End-to-End Latency----------------						
Mean E2E Latency (ms):                   21616.17  						
Median E2E Latency (ms):                 21256.22  						
P90 E2E Latency (ms):                    41068.15  						
P99 E2E Latency (ms):                    45190.67  						
---------------Time to First Token----------------						
Mean TTFT (ms):                          355.76    						
Median TTFT (ms):                        232.82    						
P99 TTFT (ms):                           1063.57   						
-----Time per Output Token (excl. 1st token)------						
Mean TPOT (ms):                          45.13     						
Median TPOT (ms):                        45.07     						
P99 TPOT (ms):                           50.06     						
---------------Inter-Token Latency----------------						
Mean ITL (ms):                           45.13     						
Median ITL (ms):                         42.94     						
P95 ITL (ms):                            52.33     						
P99 ITL (ms):                            102.35    						
Max ITL (ms):                            222.57    						
==================================================						
```

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review Process

1. Ping Merge Oncalls to start the PR flow. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - `/tag-run-ci-label`, `/rerun-failed-ci`, `/tag-and-rerun-ci`
4. After green CI and required approvals, ask Merge Oncalls to merge.

### CODE REVIEW

File: python/sglang/srt/models/mindspore.py
Comment (chz34) [inline python/sglang/srt/models/mindspore.py]: use `batch_valid_length` instead of `seq_lens`, because it's modifiied in speculative mode.

File: python/sglang/srt/models/mindspore.py [hunk 1]
```diff
@@ -237,24 +241,38 @@ def prepare_cache(cache_list, is_key_cache):
...
-        # Without any prefix cache => Use FlashAttentionScore
-        # With cache => Use PagedAttention, no matter the query length is 1 or not
-        is_prefill = forward_batch.forward_mode.is_extend()
-        is_prefill = is_prefill and forward_batch.extend_prefix_lens.sum().item() == 0
-
+        is_prefill = self.get_is_prefill(forward_batch)
         batch_valid_length = forward_batch.seq_lens.cpu().numpy()
-
+        if forward_batch.forward_mode.is_target_verify():
+            batch_valid_length += forward_batch.spec_info.num_tokens_per_batch
```
Comment (chz34) [inline python/sglang/srt/models/mindspore.py]: batch_valid_lenght is no longer equal to seq_lens,  need to avoid using seq_lens in the following code.

File: python/sglang/srt/models/mindspore.py
Comment (zzzzzzzxh) [reply python/sglang/srt/models/mindspore.py]: done

File: python/sglang/srt/model_executor/model_runner.py [hunk 2]
```diff
@@ -2353,6 +2355,40 @@ def sample(
...
+                # If it's a single integer (0-d tensor converted to list might behave differently depending on conversion),
+                # but sample usually returns 1D tensor [batch_size].
+                if isinstance(ids_list, int):
+                    ids_list = [ids_list]
+                
+                decoded_text = self.tokenizer.batch_decode(ids_list)
+                print(f"DEBUG: Decoded text: {decoded_text}", flush=True)
+            except Exception as e:
+                print(f"DEBUG: Failed to decode tokens: {e}", flush=True)
+
```
[the remark this answers is not in the document]
Comment (zzzzzzzxh) [reply python/sglang/srt/model_executor/model_runner.py]: removed

File: python/sglang/srt/speculative/eagle_worker.py [hunk 3]
```diff
@@ -296,6 +296,26 @@ def forward_batch_generation(self, batch: ScheduleBatch) -> GenerationBatchResul
...
+                        draft_ids = spec_info.draft_token
+                        if isinstance(draft_ids, torch.Tensor):
+                            ids_list = draft_ids.cpu().tolist()
+                        else:
+                            ids_list = list(draft_ids)
+                        
+                        decoded_text = tokenizer.batch_decode(ids_list)
+                        print(f"DEBUG: EagleV1 Draft Model Output (Candidates): {decoded_text}", flush=True)
+                except Exception as e:
+                    print(f"DEBUG: Failed to decode EagleV1 draft tokens: {e}", flush=True)
```
[the remark this answers is not in the document]
Comment (zzzzzzzxh) [reply python/sglang/srt/speculative/eagle_worker.py]: removed

File: python/sglang/srt/speculative/eagle_worker.py [hunk 4]
```diff
@@ -743,6 +763,21 @@ def verify(self, batch: ScheduleBatch, spec_info: EagleVerifyInput):
...
+                # res.verified_id contains the accepted token IDs
+                if isinstance(res.verified_id, torch.Tensor):
+                    ids_list = res.verified_id.cpu().tolist()
+                else:
+                    ids_list = list(res.verified_id)
+                
+                decoded_text = tokenizer.batch_decode(ids_list)
+                print(f"DEBUG: EagleV1 Target Model Output (Final Verified): {decoded_text}", flush=True)
+            except Exception as e:
+                print(f"DEBUG: Failed to decode EagleV1 tokens: {e}", flush=True)
```
[the remark this answers is not in the document]
Comment (zzzzzzzxh) [reply python/sglang/srt/speculative/eagle_worker.py]: removed

File: python/sglang/srt/speculative/eagle_worker_v2.py [hunk 5]
```diff
@@ -636,6 +636,25 @@ def forward_batch_generation(self, model_worker_batch: ModelWorkerBatch):
...
+                        draft_ids = verify_input.draft_token
+                        if isinstance(draft_ids, torch.Tensor):
+                            ids_list = draft_ids.cpu().tolist()
+                        else:
+                            ids_list = list(draft_ids)
+                        
+                        decoded_text = tokenizer.batch_decode(ids_list)
+                        print(f"DEBUG: EagleV2 Draft Model Output (Candidates): {decoded_text}", flush=True)
+                except Exception as e:
+                    print(f"DEBUG: Failed to decode EagleV2 draft tokens: {e}", flush=True)
```
[the remark this answers is not in the document]
Comment (zzzzzzzxh) [reply python/sglang/srt/speculative/eagle_worker_v2.py]: removed

File: python/sglang/srt/speculative/eagle_worker_v2.py [hunk 6]
```diff
@@ -732,6 +751,39 @@ def verify(self, batch: ModelWorkerBatch):
...
+                # If the user wants the "base model verify result", it essentially IS the accepted tokens.
+                # The 'logits_output' is from the target model verify pass.
+                # The sampling step decides which ones to keep.
+                
+                # So 'predict' contains the tokens that were accepted from the draft + 1 new token from target.
+                
+                # If we want to see exactly what the target model "verified" (accepted), it is 'predict'.
+                
+            except Exception as e:
+                print(f"DEBUG: Failed to decode EagleV2 tokens: {e}", flush=True)
```
[the remark this answers is not in the document]
Comment (zzzzzzzxh) [reply python/sglang/srt/speculative/eagle_worker_v2.py]: removed

File: python/sglang/srt/speculative/eagle_worker_v2.py [hunk 7]
```diff
@@ -250,6 +251,10 @@ def init_cuda_graphs(self):
         if self.server_args.disable_cuda_graph:
             return
 
+        if is_mindspore():
+            logger.info("Skip CUDA graph capture for MindSpore backend.")
```
Comment (chz34) [inline python/sglang/srt/speculative/eagle_worker_v2.py]: Not necessary to log info here. Just return is enough, just like skip init_device_graphs in model_runner

File: python/sglang/srt/utils/common.py [hunk 8]
```diff
@@ -162,6 +162,21 @@ def is_npu() -> bool:
     return True
 
 
+def is_mindspore() -> bool:
```
Comment (chz34) [inline python/sglang/srt/utils/common.py]: Make this lru_cache

File: python/sglang/srt/models/mindspore.py [hunk 9]
```diff
@@ -196,6 +196,10 @@ def __init__(
         )
         self.key_cache = []
         self.value_cache = []
+    
+    @property
+    def hot_token_id(self):
+        return tensor_ms2torch(getattr(self.model, "hot_token_id", None))
```
Comment (ping1jing2) [inline python/sglang/srt/models/mindspore.py]: please use `.get()` instead of `getattr`, and never use `getattr/setattr` in SGLang

File: python/sglang/srt/models/mindspore.py [hunk 10]
```diff
@@ -236,30 +240,44 @@ def prepare_cache(cache_list, is_key_cache):
 
         return mutable(self.key_cache), mutable(self.value_cache)
 
+    def get_is_prefill(self, forward_batch: ForwardBatch):
```
Comment (ping1jing2) [inline python/sglang/srt/models/mindspore.py]: this function is confused, could you please change it

File: python/sglang/srt/models/mindspore.py [hunk 11]
```diff
@@ -313,5 +340,22 @@ def get_model_config_for_expert_location(cls, config):
...
+        return tensor_ms2torch(embed), tensor_ms2torch(head)
+
+    def set_embed_and_head(self, embed, head):
+        self.model.set_embed_and_head(tensor_torch2ms(embed), tensor_torch2ms(head))
+
+    def get_embed(self):
+        return tensor_ms2torch(self.model.get_embed())
+
+    def set_embed(self, embed):
+        self.model.set_embed(tensor_torch2ms(embed))
```
Comment (ping1jing2) [inline python/sglang/srt/models/mindspore.py]: for these functions, please discuss internal to decide use `property` or `function`

File: python/sglang/srt/models/mindspore.py (hunk 10 above)
Comment (zzzzzzzxh) [reply python/sglang/srt/models/mindspore.py]: rename def _is_prefill,  return True/False

File: python/sglang/srt/models/mindspore.py (hunk 9 above)
Comment (zzzzzzzxh) [reply python/sglang/srt/models/mindspore.py]: done

**Review by ping1jing2 — APPROVED** [review]
